src/script/sign.h

```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+            }
+        }
+    }
+
+    template <typename Stream>
+    PartiallySignedInput(deserialize_type, Stream& s) {
+        Unserialize(s);
+    }
+};
+
+/** A structure for PSBTs which contains per output information */
+struct PSBTOutput
+{
+    CScript redeem_script;
+    CScript witness_script;
+    std::map<CPubKey, std::vector<uint32_t>> hd_keypaths;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+
+    bool IsNull() const;
+    PSBTOutput() {}
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        // Write the redeem script
+        if (!redeem_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_REDEEMSCRIPT);
+            s << redeem_script;
+        }
+
+        // Write the witness script
+        if (!witness_script.empty()) {
+            SerializeToVector(s, PSBT_OUT_WITNESSSCRIPT);
+            s << witness_script;
+        }
+
+        // Write any hd keypaths
+        for (auto keypath_pair : hd_keypaths) {
+            SerializeToVector(s, PSBT_OUT_BIP32_DERIVATION, MakeSpan(keypath_pair.first));
+            WriteCompactSize(s, keypath_pair.second.size() * sizeof(uint32_t));
+            for (auto& path : keypath_pair.second) {
+                s << path;
+            }
+        }
+
+        // Write unknown things
+        for (auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        s << PSBT_SEPARATOR;
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // read size of key
+            uint64_t key_len = ReadCompactSize(s);
+
+            // the key length is 0 if that was actually a separator byte
+            // This is a special case for key lengths 0 as those are not allowed (except for separator)
+            if (key_len == 0) {
+                return;
+            }
+
+            // Read key
+            std::vector<unsigned char> key;
+            key.resize(key_len);
+            s >> MakeSpan(key);
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Do stuff based on type
+            switch(type) {
+                case PSBT_OUT_REDEEMSCRIPT:
+                {
+                    if (!redeem_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, output redeemScript already provided");
+                    }
+                    s >> redeem_script;
+                    break;
+                }
+                case PSBT_OUT_WITNESSSCRIPT:
+                {
+                    if (!witness_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, output witnessScript already provided");
+                    }
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_OUT_BIP32_DERIVATION:
```

Comment:
Can you abstract out the serialization/deserialization code for derivation paths and scripts into separate functions? Otherwise it is needlessly duplicated across input and outputs.

Reply:
Done